### PR TITLE
[WIP] Support Denoising Diffusion Probabilistic Models

### PR DESCRIPTION
**Pull request by LeoXing1996**

### CODE REVIEW

File: mmgen/models/diffusions/base_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+        other repo in **MM** series, we allow the back propagation and
+        optimizer updating to directly follow the iterative training schedule
+        of DDPMs.
+        Of course, we will show that you can also move the back
+        propagation outside of this method, and then optimize the parameters
+        in the optimizer hook. But this will cause extra GPU memory cost as a
+        result of retaining computational graph. Otherwise, the training
+        schedule should be modified in the detailed implementation.
+
+        TODO: training would be supported later ~~~
```
Comment (nbei) [inline mmgen/models/diffusions/base_diffusion.py:89]: Remove `~~~`. Because this is an official release version.

File: mmgen/models/diffusions/base_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+            noise (torch.Tensor | callable | None): You can directly give a
+                batch of noise through a ``torch.Tensor`` or offer a callable
+                function to sample a batch of noise data. Otherwise, the
+                ``None`` indicates to use the default noise sampler.
+            num_batches (int, optional):  The number of batch size.
+                Defaults to 0.
+            sample_model (str, optional): The model to sample. If ``ema/orig``
+                is passed, this method would try to sample from ema (if
+                ``self.use_ema == True``) and orig model. Defaults to
+                'ema/orig'.
```
Comment (nbei) [inline mmgen/models/diffusions/base_diffusion.py:111]: Add docstring for `label`.

File: mmgen/models/diffusions/base_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+                    label=None,
+                    save_intermedia=False,
+                    timesteps_noise=None,
+                    **kwargs):
+        """DDPM sample from random noise.
+        TODO: Allow user to pass their own model, otherwise we can
+        remove argument `model`.
+
+        Args:
+            timesteps_noise (torch.Tensor, optional): If set as True, noise
```
Comment (LeoXing1996) [inline mmgen/models/diffusions/base_diffusion.py:171]: If set as True --> If passed

File: mmgen/models/diffusions/base_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+            var_pred = torch.exp(log_var_pred)
+
+        elif self.denoising_var.upper() == 'FIXED_LARGE':
+            # use betas as var
+            var_pred = self.var_to_tensor(
+                np.append(self.tilde_betas_t[1], self.betas), t, tar_shape)
+            log_var_pred = torch.log(var_pred)
+
+        elif self.denoising_var.upper() == 'FIXED_SMALL':
+            # use posteroir (tilde_betas)  as var
```
Comment (LeoXing1996) [inline mmgen/models/diffusions/base_diffusion.py:472]: posterior

File: mmgen/models/architectures/DDPM/modules.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+@ACTIVATION_LAYERS.register_module()
+class SiLU(nn.Module):
+
+    def __init__(self, inplace=False):
+        super().__init__()
+        self.inplace = inplace
+
+    def forward(self, x):
+        # return F.silu(x, inplace=self.inplace)
+        return torch.sigmoid(x) * x
```
Comment (plyfager) [inline mmgen/models/architectures/DDPM/modules.py:41]: Dose `inplace` make sense here?

File: mmgen/models/architectures/DDPM/modules.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+class DenoisingUpsample(nn.Module):
+
+    def __init__(self, in_channels, use_conv=True):
+        super().__init__()
+        if use_conv:
+            self.conv = nn.Conv2d(in_channels, in_channels, 3, 1, 1)
+
+    def forward(self, x):
+        x = F.interpolate(x, scale_factor=2, mode='nearest')
+        if hasattr(self, 'conv'):
```
Comment (plyfager) [inline mmgen/models/architectures/DDPM/modules.py:322]: I suggest using `with_conv`.

File: mmgen/models/architectures/DDPM/denoising.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                 use_scale_shift_norm=False,
+                 num_heads=4,
+                 resblock_cfg=dict(type='DenoisingResBlock'),
+                 attention_cfg=dict(type='MultiHeadAttention'),
+                 downsample_conv=True,
+                 upsample_conv=True,
+                 downsample_cfg=dict(type='DenoisingDownsample'),
+                 upsample_cfg=dict(type='DenoisingUpsample'),
+                 attention_res=[16, 8],
+                 use_checkpoint=False):
```
Comment (plyfager) [inline mmgen/models/architectures/DDPM/denoising.py:47]: As a convention, we use the argument `pretrained`. Here you may need to give further explanation.

File: mmgen/models/architectures/DDPM/denoising.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+        super().__init__()
+
+        self.num_classes = num_classes
+        self.num_timesteps = num_timesteps
+        self.rescale_timesteps = rescale_timesteps
+
+        self.output_cfg = deepcopy(output_cfg)
+        self.mean_cfg = getattr(self.output_cfg, 'mean', 'eps')
```
Comment (plyfager) [inline mmgen/models/architectures/DDPM/denoising.py:56]: You have to notice that `output_cfg` is a dict and 'mean' is not an attribute.

File: mmgen/models/architectures/DDPM/denoising.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        elif self.mean_cfg.upper() == 'PREVIOUS_X':
+            output_dict['x_tm1_pred'] = mean
+
+        if return_noise:
+            output_dict['x_t'] = x_t
+            output_dict['t'] = t
+
+        return output_dict
+
+    def init_weights(self, pretrained=None):
```
Comment (plyfager) [inline mmgen/models/architectures/DDPM/denoising.py:248]: This function is not used?

File: mmgen/models/diffusions/base_diffusion.py [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+            denoising,
+            default_args=dict(
+                num_classes=num_classes, num_timesteps=num_timesteps))
+
+        # get output-related configs from denoising
+        self.denoising_var = self.denoising.var_cfg
+        self.denoising_mean = self.denoising.mean_cfg
+        # output_channels in denoising may be double, therefore we
+        # get number of channels from config
+        image_channels = self._denoising_cfg.get('in_channels')
```
[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/diffusions/base_diffusion.py:57]: `in_channels` is a must for `denoising` model, so I not use default value here.

File: mmgen/models/architectures/DDPM/modules.py (hunk 5 above)
Comment (LeoXing1996) [reply mmgen/models/architectures/DDPM/modules.py:41]: I forget to add `TODO` to this comment.
Official code realizes `SiLU` as L:41, if we use torch's official implementation, we may have an error of 1e-5.
I'm not sure which version to use.

File: mmgen/models/architectures/DDPM/denoising.py (hunk 7 above)
Comment (LeoXing1996) [reply mmgen/models/architectures/DDPM/denoising.py:47]: Here `checkpoint` means [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html), which allow us to use larger batch size with limited GPU memory.
